test/mapdefaults/combos/both/map_test.go

```diff
@@ -0,0 +1,111 @@
+// Protocol Buffers for Go with Gadgets
+//
+// Copyright (c) 2015, The GoGo Authors. All rights reserved.
+// http://github.com/gogo/protobuf
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package mapdefaults
+
+import (
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+)
+
+func TestUnmarshalImplicitDefaultKeyValue1(t *testing.T) {
+	// The follow message is generated via python:
+	//
+	// map = MapTest()
+	// msg.str_str['foo'] = ''
+	// msg.str_str[''] = 'bar'
```

Comment:
Maybe we could generate this with here with a proto go message.

```proto
message Map {
   repeated OldMap OldMap = 1;
}

message OldMap {
   optional string key = 1;
   optional string value = 2;
}
```
```go
m := &Map{}
m.OldMap = append(m.OldMap, &OldMap{Key: "foo", Value: ""}
m.OldMap = append(m.OldMap, &OldMap{Key: "", Value: "bar"}
...
```

This way we can also test whether we are skipping field 3, by adding a field number 3 to OldMap.